lib/client.js

```diff
@@ -219,12 +220,37 @@ MqttClient.prototype._setupStream = function () {
   // Echo connection errors
   parser.on('error', this.emit.bind(this, 'error'));
 
-  this.outgoingStore
-    .createStream()
-    .on('data', function (packet) {
-      that._sendPacket(packet);
+  outStore = this.outgoingStore
+  .createStream();
+  // Control of stored messages
+  outStore.once('readable', function () {
+      function storeDeliver () {
+        var packet = outStore.read(1);
+        if (!packet) {
+          return;
+        }
+        // Avoid unnecesary stream read operations
+        if (!that.disconnecting && !that.reconnectTimer) {
+          outStore.read(0);
+          // Ensure that the next message will only be read after callback is issued
+          that.outgoing[packet.messageId] = function () {
+            storeDeliver();
+          };
+          that._sendPacket(packet);
+        }
```

Comment:
I am currently using ```(!that.disconnecting && !that.reconnectTimer && (0 < that.options.reconnectPeriod))```to check if the client disconnected but sometimes (unexpectedly) this condition becomes true (presumably ```!that.reconnectTimer``` fault) even when the client is not connected. Checking the ```this.connected```but this flag is not ```true```when the ```_setupStream``` is created when the client is first created. 
I will continue to use this condition because it reduces the Store reading calls, but any other suggestions are welcomed.